tests/Doctrine/Tests/Common/Collections/CollectionTest.php

```diff
@@ -37,6 +41,36 @@ public function testMatching(): void
         self::assertEquals(1, count($col));
     }
 
+    public function testMatchingCallable(): void
```

Comment:
Hello,

I can't understand why Criteria accepts Value object. [ArrayCollection::matching method will call array_filter with $filter](https://github.com/doctrine/collections/blob/534712ad689fed8c2609f86e6f3466fb24eb4a6d/lib/Doctrine/Common/Collections/ArrayCollection.php#L448) as $value of Value::__construct. This generates some weird behavior:
```php
$this->collection->matching(new Criteria(new Value('undeclared_func')));

// TypeError: array_filter() expects parameter 2 to be a valid callback, function 'undeclared_func' not found or invalid function name
```


```php
$this->collection->matching(new Criteria(new Value('is_string')));

// Works
```

Is this expected?